![pycoMeth](https://pmbio.github.io/pycoMeth/images/pycoMeth_long.png)

[![GitHub license](https://img.shields.io/github/license/a-slide/pycoMeth.svg)](https://github.com/a-slide/pycoMeth/blob/master/LICENSE)
[![Language](https://img.shields.io/badge/Language-Python3.7+-yellow.svg)](https://www.python.org/)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7373424.svg)](https://doi.org/10.5281/zenodo.7373424)

[![PyPI version](https://badge.fury.io/py/pycoMeth.svg)](https://badge.fury.io/py/pycoMeth)
[![PyPI downloads](https://pepy.tech/badge/pycoMeth)](https://pepy.tech/project/pycoMeth)
[![Anaconda Version](https://anaconda.org/snajder-r/pycometh/badges/version.svg)](https://anaconda.org/snajder-r/pycometh)
[![Anaconda Downloads](https://anaconda.org/snajder-r/pycometh/badges/downloads.svg)](https://anaconda.org/snajder-r/pycometh)

---
Version in this branch: 2.2.2

---

**DNA methylation analysis downstream to Nanopolish for Oxford Nanopore DNA sequencing datasets**

`pycoMeth` can be used for further analyses starting from the output files generated by [`Nanopolish call-methylation`](https://github.com/jts/nanopolish). The package contains a suite of tools to **find CpG islands**, **segment methylome**, and to perform a **differential methylation analysis** across multiple samples.

`pycoMeth` generates extensive tabulated reports and BED files which can be loaded in a genome browser. In addition, an interactive HTML report of differentially
methylated intervals/islands can also generated at the end of the analysis.

[`Methplotlib`](https://github.com/wdecoster/methplotlib) developed by [Wouter de coster](https://twitter.com/wouter_decoster) is an excellent complementary tool to visualise and explore methylation status for specific loci.

Please be aware that `pycoMeth` is a research package that is still under development. The API, command line interface, and implementation might change without retro-compatibility.

---
### Installation

Install either using conda:

    conda install -c snajder-r -c bioconda -c conda-forge pycometh

Or using pip:
    
    pip install pycometh

### Documentation

A more detailed usage documentation can be found at https://pmbio.github.io/pycoMeth/

### pycoMeth workflow

![Workflow](https://pmbio.github.io/pycoMeth/images/overview.svg)


### pycoMeth example HTML report

[Example HTML report 1](https://pmbio.github.io/pycoMeth/examples/pycoMeth_summary_report.html)

[Example HTML interval report](https://pmbio.github.io/pycoMeth/examples/interval_reports/interval_0020_chr19-44756986-44758607.html)

---

### Citing

The repository is archived at Zenodo. https://doi.org/10.5281/zenodo.6637645

If you find pycoMeth useful, please cite our preprint:

Snajder, Rene H., Oliver Stegle, and Marc Jan Bonder. 2022. "PycoMeth: A Toolbox for Differential Methylation Testing from Nanopore Methylation Calls." bioRxiv. https://doi.org/10.1101/2022.02.16.480699.


    @article {Snajder2022.02.16.480699,
        author = {Snajder, Rene and Leger, Adrien and Stegle, Oliver and Bonder, Marc Jan},
	    title = {pycoMeth: A toolbox for differential methylation testing from Nanopore methylation calls},
	    year = {2022}, doi = {10.1101/2022.02.16.480699}, publisher = {Cold Spring Harbor Laboratory},
        journal = {bioRxiv}
    }


### Authors

* Rene Snajder (rene.snajder@gmail.com)
* Adrien Leger
